fs/diriterate/diriterate.go

```diff
@@ -24,28 +24,40 @@ import (
 	scalibrfs "github.com/google/osv-scalibr/fs"
 )
 
+// readDirAll reads all directory entries using fsys.ReadDir and returns a DirIterator with preloaded entries.
+func readDirAll(fsys scalibrfs.FS, name string) (*DirIterator, error) {
+	files, err := fsys.ReadDir(name)
+	if err != nil {
+		return nil, &fs.PathError{Op: "readdir", Path: name, Err: errors.New("not implemented")}
+	}
+	return &DirIterator{files: files, curr: 0}, nil
+}
+
 // ReadDir reads the named directory and returns an iterator over the directory entries.
 func ReadDir(fsys scalibrfs.FS, name string) (*DirIterator, error) {
+	// Check if the path is accessible
+	_, err := fsys.Stat(name)
+	if err != nil {
+		return nil, &fs.PathError{Op: "stat", Path: name, Err: err}
+	}
+
+	// Try to open the directory
 	file, err := fsys.Open(name)
 	if err != nil {
-		return nil, err
+		// If Open fails, fall back to reading all entries
```

Comment:
Could be reworded to clarify when this can happen:

"The underlying filesystem might not have implemented Open() for directories. In this case, we fall back to using ReadDir()"